streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java

```diff
@@ -1142,7 +1147,7 @@ void retryWithBackoff(final Map<TaskId, Set<TopicPartition>> tasksToBeCreated) {
                     }
                 }
 
-                if (tasksToBeCreated.isEmpty()) {
+                if (tasksToBeCreated.isEmpty() || backoffTimeMs > rebalanceTimeoutMs) {
```

Comment:
I'm not sure this is correct. Lets say we have 2 tasks. Currently,  once the first task is successful we will sleep for `backoffTimeMs`. This shouldn't happen, i.e., once the task was successful `backoffTimeMs` should be set back to the initial value and we shouldn't sleep before we attempt to create the next task.

We need to track how long we've been rebalancing in total and use that to determine if we should give up (rather than the `backoffTimeMs`

In the event that it times out i think we should probably throw an exception so we leave the group, the thread shuts down (releasing any acquired locks), and some other thread can pick up the work.